File: packages/shared/src/douyin_engagement_zero_sentinels.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Literal

DouyinEngagementMetric = Literal["comment", "share", "like", "view"]

COMMENT_ZERO_SENTINELS: tuple[str, ...] = (
    "抢首评",
    "快来抢首评",
    "抢沙发",
)

SHARE_ZERO_SENTINEL_PATTERN = re.compile(r"^分享$")


@dataclass(frozen=True)
class DouyinEngagementParseResult:
    kind: Literal["numeric", "zero_sentinel", "missing"]
    value: int | None = None
    raw_text: str | None = None
    sentinel: str | None = None


def _compact_text(value: Any) -> str:
    if not isinstance(value, str):
        return ""
    return re.sub(r"\s+", " ", value.strip())


def parse_douyin_engagement_text(
    metric: DouyinEngagementMetric,
    raw_text: Any,
    *,
    share_icon_context: bool = False,
) -> DouyinEngagementParseResult:
    text = _compact_text(raw_text)
    if not text:
        return DouyinEngagementParseResult(kind="missing")

    if metric == "comment":
        for sentinel in COMMENT_ZERO_SENTINELS:
            if text == sentinel or sentinel in text:
                return DouyinEngagementParseResult(kind="zero_sentinel", value=0, raw_text=sentinel, sentinel=sentinel)
        return DouyinEngagementParseResult(kind="missing")

    if metric == "share":
        if share_icon_context and SHARE_ZERO_SENTINEL_PATTERN.fullmatch(text):
            return DouyinEngagementParseResult(kind="zero_sentinel", value=0, raw_text=text, sentinel=text)
        return DouyinEngagementParseResult(kind="missing")

    return DouyinEngagementParseResult(kind="missing")


def normalize_douyin_engagement_count(
    metric: DouyinEngagementMetric,
    value: Any = None,
    text: Any = None,
    *,
    share_icon_context: bool = False,
) -> int | None:
    if value is not None and value != "":
        if isinstance(value, (int, float)):
            parsed = int(float(value))
            return parsed if parsed >= 0 else None
    compact = _compact_text(text)
    if not compact:
        return None
    parsed = parse_douyin_engagement_text(metric, compact, share_icon_context=share_icon_context)
    if parsed.kind == "zero_sentinel":
        return 0
    return None

File: apps/extension-douyin-capture/src/douyinEngagementZeroSentinels.ts
export type DouyinEngagementMetric = "comment" | "share" | "like" | "view";

export type DouyinEngagementParseResult =
  | { kind: "numeric"; value: number; rawText: string }
  | { kind: "zero_sentinel"; value: 0; rawText: string; sentinel: string }
  | { kind: "missing" };

export const COMMENT_ZERO_SENTINELS = ["抢首评", "快来抢首评", "抢沙发"] as const;

const SHARE_ZERO_SENTINEL_PATTERN = /^分享$/;

function compactText(rawText: string | null | undefined): string {
  if (!rawText) return "";
  return rawText.replace(/\s+/g, " ").trim();
}

export function parseDouyinEngagementText(
  metric: DouyinEngagementMetric,
  rawText: string | null | undefined,
  options?: { shareIconContext?: boolean }
): DouyinEngagementParseResult {
  const text = compactText(rawText);
  if (!text) return { kind: "missing" };

  if (metric === "comment") {
    for (const sentinel of COMMENT_ZERO_SENTINELS) {
      if (text === sentinel || text.includes(sentinel)) {
        return { kind: "zero_sentinel", value: 0, rawText: sentinel, sentinel };
      }
    }
    return { kind: "missing" };
  }

  if (metric === "share") {
    if (options?.shareIconContext && SHARE_ZERO_SENTINEL_PATTERN.test(text)) {
      return { kind: "zero_sentinel", value: 0, rawText: text, sentinel: text };
    }
    return { kind: "missing" };
  }

  return { kind: "missing" };
}

export function parseDouyinEngagementCount(
  metric: DouyinEngagementMetric,
  rawText: string | null | undefined,
  options?: { shareIconContext?: boolean }
): number | null {
  const parsed = parseDouyinEngagementText(metric, rawText, options);
  return parsed.kind === "zero_sentinel" ? 0 : null;
}
